src/nvim/eval.c

```diff
@@ -21757,3 +21971,81 @@ bool eval_has_provider(char *name)
 
   return false;
 }
+
+static DictWatcher *dictwatcher_node_data(QUEUE *q)
+{
+  return QUEUE_DATA(q, DictWatcher, node);
+}
+
+static void dictwatcher_notify(dict_T *dict, const char *key,
+    typval_T *newtv, typval_T *oldtv)
+{
+  typval_T argv[2];
+  for (size_t i = 0; i < ARRAY_SIZE(argv); i++) {
+    init_tv(argv + i);
+  }
+
+  argv[0].v_type = VAR_STRING;
+  argv[0].vval.v_string = (uint8_t *)xstrdup(key);
+  argv[1].v_type = VAR_DICT;
+  argv[1].vval.v_dict = dict_alloc();
+  argv[1].vval.v_dict->dv_refcount++;
+
+  if (newtv) {
+    dictitem_T *v = dictitem_alloc((uint8_t *)"new");
+    copy_tv(newtv, &v->di_tv);
+    dict_add(argv[1].vval.v_dict, v);
+  }
+
+  if (oldtv) {
+    dictitem_T *v = dictitem_alloc((uint8_t *)"old");
+    copy_tv(oldtv, &v->di_tv);
+    dict_add(argv[1].vval.v_dict, v);
+  }
+
+  typval_T rettv;
+
+  QUEUE *w;
+  QUEUE_FOREACH(w, &dict->watchers) {
```

Comment:
Creating a DFA is much harder then having two functions and using nested dictionaries. Also watching `g:` is not needed because watchers are attached to dictionaries and not to keys: specifically I was thinking that you *do* notify about additions and removals and in this case `watcheradd(dict, "non_existing_key", watcher)` is going to call watcher when this key is added and watching `g:` is not necessary.